hass_nabucasa/llm.py

```diff
@@ -0,0 +1,293 @@
+    """Normalized LLM image generation response."""
+
+    mime_type: str
+    image_data: bytes
+    model: str | None
+    width: int | None
+    height: int | None
+    revised_prompt: str | None
+
+
+class LLMImageAttachment(TypedDict):
+    """Image attachment for editing requests."""
+
+    filename: str | None
+    mime_type: str | None
+    data: bytes
+
+
+class LLMRequestError(LLMError):
+    """Base error for LLM generation failures."""
+
+
+class LLMAuthenticationError(LLMRequestError):
+    """Raised when LLM authentication fails."""
+
+
+class LLMRateLimitError(LLMRequestError):
+    """Raised when LLM requests are rate limited."""
+
+
+class LLMServiceError(LLMRequestError):
+    """Raised when LLM requests fail due to service issues."""
+
+
+class LLMResponseError(LLMRequestError):
+    """Raised when LLM responses are unexpected."""
+
+
+@api_exception_handler(LLMServiceError)
+async def _async_fetch_image_from_url(url: str) -> bytes:
+    """Fetch image data from a URL asynchronously."""
+    async with aiohttp.ClientSession() as session, session.get(url) as response:
+        response.raise_for_status()
+        return await response.read()
+
+
+@api_exception_handler(LLMServiceError)
+async def _extract_response_image_data(
+    response: dict[str, Any],
+) -> LLMGeneratedImage:
+    data = response.get("data")
+    if not data or not isinstance(data, list):
+        raise LLMResponseError("Unexpected response from Cloud LLM")
+
+    item = data[0]
+
+    url = item.get("url")
+    b64 = item.get("b64_json")
+
+    if not b64 and url:
+        image_bytes = await _async_fetch_image_from_url(url)
+        b64 = base64.b64encode(image_bytes).decode("utf-8")
+
+    if not b64:
+        raise ValueError(
+            "Image generation response contains neither url nor b64_json.")
+
+    decoded_image = base64.b64decode(b64)
+
+    return LLMGeneratedImage(
+        mime_type="image/png",
+        model=response.get("model"),
+        image_data=decoded_image,
+        width=item.get("width"),
+        height=item.get("height"),
+        revised_prompt=item.get("revised_prompt"),
+    )
+
+
+class LLM(ApiBase):
+    """Class to handle LLM services."""
+
+    def __init__(self, cloud: Cloud[_ClientT]) -> None:
+        """Initialize LLM services."""
+        super().__init__(cloud)
+        self._token: str | None = None
+        self._base_url: str | None = None
+        self._generate_data_model: str | None = None
+        self._generate_image_model: str | None = None
+        self._conversation_model: str | None = None
+        self._valid_until: datetime | None = None
+        self._lock = asyncio.Lock()
+
+    def _validate_token(self) -> bool:
+        """Validate token outside of coroutine."""
+        # Add a 5-minute buffer to avoid race conditions near expiry
+        return self._cloud.valid_subscription and bool(
+            self._valid_until and utcnow() + timedelta(minutes=5) < self._valid_until
```

Comment:
Store the `timedelta(minutes=5)` as a module const.

Example: `TOKEN_EXP_BUFFER_MINUTES = timedelta(minutes=5)`